Add missing header in socket_wrapper.cpp

alinbalutoiu opened the pull request:
Inside the socket_wrapper.cpp file there is a call to strerror_r
which raises an error. As a workaround, the fpermissive flag can
be used which will allow the code to compile even if it doesn't
conform the language rules.

The fpermissive compilation flag can be avoided by including string.h.

msftclas commented:
@alinbalutoiu,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by Microsoft. We will now review your pull request.
_Thanks,
Microsoft Pull Request Bot_

EMumau reviewed (changes requested):
By chance, do you mean to include the string header file from the C standard libraries, or do you indeed mean to include a header file name string.h from the include path?
#include <cstring>

alinbalutoiu commented:
I meant to include the string.h from the C Standard Library files which is required for strerror_r (https://linux.die.net/man/3/strerror_r). The function strerror_r is also used inside "server.hpp" but we have there "MI.h" included which also includes string.h (https://github.com/Microsoft/omi/blob/master/Unix/common/MI.h#L17). That's why there is no error raised for "server.hpp".

However, I just noticed that OMI handles differently the case when strerror_r is supported or not (https://github.com/Microsoft/omi/blob/master/Unix/configure#L1298-L1336) and threats both cases (https://github.com/Microsoft/omi/blob/master/Unix/omi_error/omierror.c#L352-L495).
Should we do the same in the script provider?

EMumau commented:
I guess that my point was missed.  My point was:
    #include <string.h> // is for C
    #include <cstring>  // is for C++

After looking into omierror.c, we should probably replicate that behavior to maintain portability.  The man page for strerr_r indicates that the behavior is potentially different on different platforms.  I did not account for that.  Please go ahead and change the definition in server.hpp to match the behavior in omierror.c.  Then in socket_wrapper.cpp, remove the definitions from the unnamed namespace and include server.hpp.

alinbalutoiu committed:
Fix strerror_r usage in socket_wrapper.cpp

Inside the socket_wrapper.cpp file there is a call to strerror_r
which raises an error. As a workaround, the fpermissive flag can
be used which will allow the code to compile even if it doesn't
conform the language rules.

The fpermissive compilation flag can be avoided by removing the
definitions from the unnamed namespace and including server.hpp
in socket_wrapper.cpp.

This patch also changes the usage of strerror_r to take into
account the XSI-compliant version.
